Multi GPU support

davidwyld opened the pull request:
Hey @fredblain I've reached the limit of what I can test on a single GPU machine with this. Would you like to give it a test on a multi-GPU system or I guess you could allow me to use your server.

I'm not 100% sure the lines I've added to the models are in precisely the right place, could you take a look and see if you agree?

Changes made to `__main__.py` add the command line option to set the gpu ID. For using deepquest as a python module, this is probably best done in the user's python code before importing deepquest (similarly to the seed setting).

Also, we can deal with this some time in the future if you prefer.

Commit:
add n_gpus training parameter

Commit:
add multi gpu functionality to models

Commit:
add multi gpu functionality to models

davidwyld commented:
(copied chat from https://github.com/RSE-Sheffield/deepquest/issues/38#issuecomment-590267932)
This draft PR facilitates setting the GPU id via `dq train -c blah.yml gpuid 0` for single GPU or `dq train -c blah.yml gpuid 0,1` for multi GPU, provided that the number of GPUs specified on the cmd line match N_GPUS in the config (eg in this case `N_GPUS: 2`)

However, this is just at the stage where it needs to be tested on a multi GPU machine.

Regarding using as a python module, the user will just have to set `os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"` and `os.environ["CUDA_VISIBLE_DEVICES"]` in python before importing deepquest as these environment variables need to be set beforehand.

Sounds like we also need tensorflow-gpu added to the requirements as well (for TF<2.0 at least)

fredblain commented on deepquest/__main__.py:
```diff
@@ -33,6 +37,17 @@ def score(args):
+    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+    gpustr = ''
+    for g in gpuid:
+        gpustr += str(g) + ','
```
It is more intuitive to write
`gpu-id: 1,2` 
than 
`gpu-id: 1 2`
in the config file, but since both could happen, I would consider both cases with something like:
```
gpuid = gpuid.split(',') if ',' in gpuid else gpuid.split()    
```
BTW, this also works if there is more than 10 GPUs (i.e. gpuid > 9, with `gpuid = '10'` resulting into two GPU IDs of 1 and 0 in the current version of the code.

fredblain commented on deepquest/__main__.py:
```diff
@@ -33,6 +37,17 @@ def score(args):
+    os.environ["CUDA_VISIBLE_DEVICES"] = gpustr[0:-1]
+    print(os.environ["CUDA_VISIBLE_DEVICES"])
+    import sys
+    sys.exit()
```
I don't get these two lines, what are they for in this function?

davidwyld replied:
Woops these are left in from debugging, will remove. Thanks.

fredblain commented on deepquest/qe_models/birnn_word.py:
```diff
@@ -139,4 +140,9 @@ def build(self):
                 outputs=[output_qe_layer]
                 )
+        
+        if params.get('N_GPUS', 1) > 1:
```
How does this work? `N_GPUS` isn't set at the moment, is it?
I guess it should be in the `set_gpu_id` function.

davidwyld replied:
Yes, in this branch it's set in the config file, but you're right, we should set it automatically.

fredblain reviewed (dismissed):
I've commented on a few points as I'm not sure the code would work as is.

davidwyld committed:
handle gpuid arguments with commas, spaces or >1 digit

davidwyld committed:
set N_GPUS and GPU_ID parameters in __main__

davidwyld committed:
add tensorflow-gpu to requirements

Commit:
fix bug in model for multi gpu

Commit:
allow GPU_ID to be set in config file

davidwyld commented:
This now works with `GPU_ID` set in a config file as eg `GPU_ID: 2,3` or set on the command line as `dq train -c config.yml changes GPU_ID=2,3` (although this will change to `--changes`) on merge with the `hotfix-predict` branch.

Limitations of this "working" are (as discussed here https://github.com/RSE-Sheffield/deepquest/issues/38):
- It will train a multi-gpu model, on the right GPUs
- but, it seems that it doesn't fully utilise all the GPUs

We should go ahead and merge this anyway, in order to clean up the tree.

fredblain commented on deepquest/__main__.py:
```diff
@@ -67,7 +69,10 @@ def changes2dict(args):
                 else:
                     try:
-                        changes[k] = ast.literal_eval(v)
+                        if k == 'GPU_ID':
+                            changes[k] = str(v)
+                        else:
+                            changes[k] = ast.literal_eval(v)
```
What prevent `ast.literal_eval()` from being used for GPU_ID as well?

fredblain replied:
> This may have to be a temporary fix, but I realised that if GPU_ID is set
to something like `2,3` (ie with a comma) then `ast.literal_eval()` turns
it into a tuple, which causes problems later down the line.

If it is temporary, what would be the other solutions?

davidwyld replied:
I'm not sure, perhaps @drj11 might have a suggestion.

drj11 replied:
Not really, other than to perhaps reverse the presumption of using `ast.literal_eval`.

`ast.literal_eval` is useful if you want something like a list or a complex number in your config values, but maybe we only need that for a few things, and we should use it only where we need it, instead of everywhere except `GPU_ID`?

fredblain reviewed (commented):
Just a quick question below regarding the use of `ast.literal_eval()`.
Otherwise, on the surface, this looks OK to me.

davidwyld commented:
This may have to be a temporary fix, but I realised that if GPU_ID is set
to something like `2,3` (ie with a comma) then `ast.literal_eval()` turns
it into a tuple, which causes problems later down the line.

--
*Dr David Wilby*
Research Software Engineer | RSE Team | University of Sheffield
d.wilby@sheffield.ac.uk | CG018, Regent Court, Portobello
rse.sheffield.ac.uk


On Wed, 1 Apr 2020 at 16:49, Fred Blain <notifications@github.com> wrote:

> *@fredblain* commented on this pull request.
>
> Just a quick question below regarding the use of ast.literal_eval().
> Otherwise, on the surface, this looks OK to me.
> ------------------------------
>
> In deepquest/__main__.py
> <https://github.com/RSE-Sheffield/nmt-keras/pull/35#discussion_r401718799>
> :
>
> > +                        if k == 'GPU_ID':
> +                            changes[k] = str(v)
> +                        else:
> +                            changes[k] = ast.literal_eval(v)
>
> What prevent ast.literal_eval() from being used for GPU_ID as well?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RSE-Sheffield/nmt-keras/pull/35#pullrequestreview-385716441>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF4277BEKUKE7UY4AFHUPRDRKNPBLANCNFSM4KVJXJIA>
> .
>

fredblain commented on deepquest/__main__.py:
```diff
@@ -33,6 +40,19 @@ def score(args):
     from . import score
     score(args.files)
 
+def set_gpu_id(gpuid):
+    import os
+    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+    gpuid = gpuid.split(',') if ',' in gpuid else gpuid.split()
+    gpustr = ''
+    gpucount = 0
+    for g in gpuid:
+        gpustr += str(g).strip() + ','
+        gpucount += 1
+    gpustr = gpustr[0:-1]
+    os.environ["CUDA_VISIBLE_DEVICES"] = gpustr
+
+    return gpucount
```
Similary to `change2dict()` that was moved to `utils/utils.py`, it is probably cleaner to move this function under `utils` as well, w.r.t. to the main functions `train()`, `predict()` and `score()`.

davidwyld replied:
Moved to `__init__.py` rather than `utils/utils.py` as `set_gpu_id()` has to happen before `import deepquest`. Commit 7cb09bd

fredblain reviewed (changes requested):
`train()`, `predict()` and `train()` are in `__main__.py` to be called in the commandline. I would rather keep it clean and move all the other functions used here, such as `set_gpu_id()`, to `deepquest/utils/utils.py`

davidwyld commented:
Yes, this will happen when the branch we made that change on is merged.

--
*Dr David Wilby*
Research Software Engineer | RSE Team | University of Sheffield
d.wilby@sheffield.ac.uk | CG018, Regent Court, Portobello
rse.sheffield.ac.uk


On Wed, 1 Apr 2020 at 17:37, Fred Blain <notifications@github.com> wrote:

> *@fredblain* requested changes on this pull request.
>
> train(), predict() and train() are in __main__.py to be called in the
> commandline. I would rather keep it clean and move all the other functions
> used here, such as set_gpu_id(), to deepquest/utils/utils.py
> ------------------------------
>
> In deepquest/__main__.py
> <https://github.com/RSE-Sheffield/nmt-keras/pull/35#discussion_r401750889>
> :
>
> > +def set_gpu_id(gpuid):
> +    import os
> +    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
> +    gpuid = gpuid.split(',') if ',' in gpuid else gpuid.split()
> +    gpustr = ''
> +    gpucount = 0
> +    for g in gpuid:
> +        gpustr += str(g).strip() + ','
> +        gpucount += 1
> +    gpustr = gpustr[0:-1]
> +    os.environ["CUDA_VISIBLE_DEVICES"] = gpustr
> +
> +    return gpucount
>
> Similary to change2dict() that was moved to utils/utils.py, it is
> probably cleaner to move this function under utils as well, w.r.t. to the
> main functions train(), predict() and score().
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RSE-Sheffield/nmt-keras/pull/35#pullrequestreview-385755993>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF4277GZWJ7LNXRV3GHY3L3RKNUVPANCNFSM4KVJXJIA>
> .
>

drj11 commented on deepquest/__main__.py:
```diff
@@ -33,6 +40,19 @@ def score(args):
+def set_gpu_id(gpuid):
+    import os
+    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+    gpuid = gpuid.split(',') if ',' in gpuid else gpuid.split()
+    gpustr = ''
+    gpucount = 0
+    for g in gpuid:
+        gpustr += str(g).strip() + ','
+        gpucount += 1
+    gpustr = gpustr[0:-1]
```
You can avoid the `if` by using an `re` to split on anything that's not a word:
    
    gpu_list = re.split(r"\W+", gpuid)

So this would allow both `0,2,3` and `0 2 3` as well as more weird things like `0 2#3`.

Your `for` loop is the same as `",".join`:

    gpustr = ",".join(gpu_list)
    gpucount = len(gpu_list)

davidwyld replied:
4443cbc will do this. I like this.

Commit:
Merge branch 'EncDoc' into deepquest-dev

Commit:
Merge branch 'deepquest-dev' of github.com:RSE-Sheffield/nmt-keras into deepquest-dev

Commit:
fix rebase conflicts

Commit:
move gpu parameter setting to __init__

Commit:
use regex to split gpu id string and simplify env var string building

Commit:
re-insert changes handling for GPU_ID

Commit:
move parameters handling functions to deepquest.__init__

this allows them to be executed by __init__ without importing any deepquest modules
to allow set_gpu_id to run before import

davidwyld commented:
@fredblain At this point (f5d7fd2), I have moved `setparameters`, `changes2dict`, `default_params` and `add_dependent_params` to `deepquest/__init__.py`. This is required by the `set_gpu_id` function if we have `GPU_ID` as a config file parameter because CUDA_VISIBLE_DEVICES needs to be set before importing any deepQuest modules (including via `from deepquest.utils.utils import setparameters` or similar)

fredblain commented:
@davidwilby, why has this branch been merged even though I expressed concerns regarding the implemented solution?

davidwyld commented:
As it is, this works correctly, this is simply a matter of elegance, which can be resolved later more simply than waiting now.
